### ISSUE TITLE
REPL history navigation should not mutate history by default

### ISSUE THREAD

**Issue by adienes** [JuliaLang/julia#53364]
example pasted from slack:

```
julia> a = "PREVIOUS"
"PREVIOUS"

julia> a = 1
1

julia> <UP><DEL><DEL><UP><CTRL-C>

julia> <UP>

# navigates to `a =` but should be `a = 1`
```

description of mechanics from @rfourquet 
> Actually, what is mutated doesn't seem to be very predictable, it looks like the mutated buffer will generally be the new REPL entry, and the history itself is not mutated, except when in the mean time you navigated somewhere else in history 

proposed change:

the in-memory representation of Julia file should _not_ be mutated during any normal navigation, except when such mutation is explicitly opted into via some new keychord

**Comment by mnemnion** [comment]
I built a REPL for Lua at one point, and decided this was the better behavior. It's quite frustrating to be editing a multi-line function, accidentally hit up arrow one time too many, then hit down arrow and find all your changes gone. Maybe `^C` should revert this? A command you have to issue to edit the past seems too modal to me.

**Comment by mbauman** [comment]
Yeah, this is very intentional (https://github.com/JuliaLang/julia/issues/15326) and I remember appreciating that very fix. But I still feel like my edits are _too_ persistent at times. I definitely want both behaviors, even though they’re obviously in conflict.

**Comment by adienes** [comment]
at the very least, when navigation is terminated via CTRL-C then history should not be committed (only when terminated via ENTER)
